File: 2015/cpp/day2/elves.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int computePaper(int a, int b, int c) {
    // surface area of the box, which is 2*l*w + 2*w*h + 2*h*l
    vector<int> sides;
    sides.push_back(a * b);
    sides.push_back(b * c);
    sides.push_back(c * a);
    int surfaceArea = 2 * (sides[0] + sides[1] + sides[2]);
    // the area of the smallest side
    // need to derefence
    int minArea = *min_element(sides.begin(), sides.end());
    return surfaceArea + minArea;
}

int computeRibbon(int a, int b, int c) {
    // shortest distance around its sides, or the smallest perimeter of any one face
    vector<int> perimeters;
    perimeters.push_back(2 * (a + b));
    perimeters.push_back(2 * (a + c));
    perimeters.push_back(2 * (b + c));
    int smallestSide = *min_element(perimeters.begin(), perimeters.end());
    // cubic feet of volume of the present
    int bow = a * b * c;
    return smallestSide + bow;
}

int main() {
    string filename = "input.txt";  // Input file name
    ifstream file(filename);       // Open the file

    if (!file.is_open()) {
        cerr << "Error opening file: " << filename << endl;
        return 1;  // Exit if file couldn't be opened
    }

    string line;
    int gifts = 0;
    int totalPaper = 0;
    int totalRibbon = 0;

    while (getline(file, line)) {
        stringstream ss(line);
        string part;
        vector<int> dims;

        // Read and parse the line into three integers
        while (getline(ss, part, 'x')) {
            dims.push_back(stoi(part));  // Convert string to integer and store in vector
        }

        // Ensure that the line contains exactly 3 integers
        if (dims.size() == 3) {
            gifts += 1;
            int a = dims[0];
            int b = dims[1];
            int c = dims[2];

            // Process the integers with the two functions
            totalPaper += computePaper(a, b, c);
            totalRibbon += computeRibbon(a, b, c);
        } else {
            cerr << "Invalid line format: " << line << endl;
        }
    }

    // Close the file
    file.close();

    // Print the accumulated results
    cout << "Number of Gifts: " << gifts << endl;
    cout << "Total Paper: " << totalPaper << endl;
    cout << "Total Robbon: " << totalRibbon << endl;

    return 0;
}

File: 2015/java/day2/Elves.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.Arrays;

public class Elves {
    public static void main(String[] args) {
        String fileName = "input.txt";
        int totalPaper = 0;
        int totalRibbon = 0;
        int i = 0;

        try (BufferedReader br = new BufferedReader(new FileReader(fileName))) {
            String line;
            while ((line = br.readLine()) != null) {
                i += 1;
                String[] parts = line.trim().split("x");
                if (parts.length == 3) {
                    try {
                        int[] dims = Arrays.stream(parts)
                            .mapToInt(Integer::parseInt)
                            .toArray();
                        int x1 = dims[0], x2 = dims[1], x3 = dims[2];

                        // total paper
                        int[] sides = {x1 * x2, x1 * x3, x2 * x3};
                        int surfaceArea = 2 * (sides[0] + sides[1] + sides[2]);
                        int minimum = Arrays.stream(sides).min().orElseThrow(() -> new IllegalArgumentException("sides array cannot be empty"));
                        totalPaper += surfaceArea + minimum;

                        // total ribbon
                        int[] perimeters = {
                            2 * x1 + 2 * x2,
                            2 * x1 + 2 * x3,
                            2 * x2 + 2 * x3
                        };
                        int smallPerim = Arrays.stream(perimeters).min().orElseThrow(() -> new IllegalArgumentException("perimeters array cannot be empty"));
                        int bow = x1 * x2 * x3;
                        totalRibbon += smallPerim + bow;

                    } catch (NumberFormatException e) {
                        System.out.println("Invalid integers found in line: " + line);
                    }
                } else {
                    System.out.println("Line does not contain exactly 3 numbers separated by 'x': " + line);
                }
            }
            System.out.println("total paper needed for " + i + " gifts: " + totalPaper + " ft^3");
            System.out.println("total ribbon needed: " + totalRibbon + " ft");
        } catch (IOException e) {
            // Handle errors related to file reading
            System.out.println("An error occurred while reading the file: " + e.getMessage());
        }
    }
}
